File: lintcode/428.cpp
class Solution {
public:
    /**
     * @param x: the base number
     * @param n: the power number
     * @return: the result
     */
    double myPow(double x, long long n) {
        // write your code here
        if (n < 0) { x = 1/x; n = -n; }
        if (n == 0) { return 1; }
        double ans = myPow(x, n / 2);
        if (n % 2 == 1) { return ans * ans * x; }
        return ans * ans;
    }
};

File: lintcode/428.py
class Solution:
    """
    @param x {float}: the base number
    @param n {int}: the power number
    @return {float}: the result
    """
    def myPow(self, x, n):
        # write your code here
        if n < 0:
            x = 1 / x
            n = -n
        if n == 0:
            return 1


        temp = self.myPow(x, n // 2) #递归深度 log(n), 时间复杂度log(n)
        if n % 2 == 0:
            return temp * temp
        else:
            return temp * temp * x
        
